granulate_utils/metrics/yarn/utils.py

```diff
@@ -0,0 +1,273 @@
+    resource_manager_index: Optional[int] = None
+
+    @cached_property
+    def is_resource_manager(self) -> bool:
+        return self.resource_manager_index is not None
+
+    @cached_property
+    def is_first_resource_manager(self) -> bool:
+        return self.resource_manager_index == 0
+
+
+def get_yarn_node_info(
+    *, logger: Union[logging.Logger, logging.LoggerAdapter], yarn_config: Optional[Dict[str, str]] = None
+) -> Optional[YarnNodeInfo]:
+    """
+    If running on YARN return YARN node information
+    """
+    config = detect_yarn_config(logger=logger) if yarn_config is None else yarn_config
+    if config is None:
+        return None
+    if rm_addresses := get_resource_manager_addresses(config, logger=logger):
+        return YarnNodeInfo(
+            resource_manager_index=get_rm_index(rm_addresses, logger=logger),
+            resource_manager_webapp_addresses=rm_addresses,
+            config=config,
+        )
+    return None
+
+
+def get_rm_index(
+    rm_addresses: List[str],
+    *,
+    logger: Union[logging.Logger, logging.LoggerAdapter],
+    hostname: Optional[str] = None,
+    ip: Optional[str] = None,
+) -> Optional[int]:
+    """
+    Return ResourceManager index for given hostname or ip
+
+    If hostname or ip are not provided, local hostname and ip are used.
+    """
+    _hostname = hostname or os.uname()[1]
+    _ip = ip or _get_local_ip(logger=logger)
+    for i, address in enumerate(rm_addresses):
+        rm_host = address.rsplit(":", 1)[0]
+        if rm_host in ("0.0.0.0", _hostname, _ip) or rm_host.startswith(f"{_hostname}."):
+            return i
+    return None
+
+
+def get_resource_manager_addresses(
+    yarn_config: Dict[str, str], *, logger: Union[logging.Logger, logging.LoggerAdapter]
+) -> Optional[List[str]]:
+    """
+    Return all ResourceManager addresses from YARN config
+
+    If high-availability mode is disabled always returns single address.
+    """
+    try:
+        config = {**RM_DEFAULTS, **yarn_config}
+        # multiple RMs in high-availability mode
+        if config.get(RM_HIGH_AVAILABILITY_ENABLED_PROPERTY_KEY) == "true":
+            logger.debug("high availability enabled, looking for RM addresses")
+            return get_all_rm_addresses(config, logger=logger)
+        # single RM
+        elif rm_address := config.get(RM_WEB_ADDRESS_PROPERTY_KEY):
+            return [resolve_variables(config, rm_address)]
+    except YarnConfigError as e:
+        logger.error("YARN config error", extra={"error": str(e)})
+    return None
+
+
+def detect_yarn_config(*, logger: Union[logging.Logger, logging.LoggerAdapter]) -> Optional[Dict[str, str]]:
+    """
+    Look for yarn-site.xml and return YARN config when found
+    """
+    if yarn_home_dir := find_yarn_home_dir(logger=logger):
+        logger.debug(f"found YARN home dir: {yarn_home_dir}")
+        yarn_site_xml_file = Path(yarn_home_dir).joinpath("./etc/hadoop/yarn-site.xml")
```

Comment:
`"./etc/hadoop/yarn-site.xml"` should be a constant

Reply:
fixed